pulpcore/plugin/stages/content_stages.py

```diff
@@ -89,43 +96,62 @@ async def run(self):
                             try:
-                                d_content.content = d_content.content.__class__.objects.get(
-                                    d_content.content.q()
+                                with transaction.atomic():
+                                    d_content.content.save()
+                            except IntegrityError as e:
+                                try:
+                                    d_content.content = d_content.content.__class__.objects.get(
+                                        d_content.content.q()
+                                    )
+                                except ObjectDoesNotExist:
+                                    raise e
+                                continue
+                            for d_artifact in d_content.d_artifacts:
+                                if not d_artifact.artifact._state.adding:
+                                    artifact = d_artifact.artifact
+                                else:
+                                    # set to None for on-demand synced artifacts
+                                    artifact = None
+                                content_artifact = ContentArtifact(
+                                    content=d_content.content,
+                                    artifact=artifact,
+                                    relative_path=d_artifact.relative_path,
                                 )
-                            except ObjectDoesNotExist:
-                                raise e
-                            continue
-                        for d_artifact in d_content.d_artifacts:
-                            if not d_artifact.artifact._state.adding:
-                                artifact = d_artifact.artifact
-                            else:
-                                # set to None for on-demand synced artifacts
-                                artifact = None
-                            content_artifact = ContentArtifact(
-                                content=d_content.content,
-                                artifact=artifact,
-                                relative_path=d_artifact.relative_path,
-                            )
-                            content_artifact_bulk.append(content_artifact)
-                ContentArtifact.objects.bulk_get_or_create(content_artifact_bulk)
+                                content_artifact_bulk.append(content_artifact)
+                    ContentArtifact.objects.bulk_get_or_create(content_artifact_bulk)
+                    if not inspect.iscoroutinefunction(self._post_save):
+                        self._post_save(batch)
+
+            if inspect.iscoroutinefunction(self._pre_save):
+                deprecation_logger.warning(
+                    "ContentSaver._pre_save() coroutine has been deprecated. Support will be "
+                    "dropped in pulpcore 3.16. ContentSaver._pre_save() is now a synchronous "
```

Comment:
```suggestion
                    "dropped in pulpcore 3.16. ContentSaver._pre_save() should now be a synchronous "
```
Also don't we need to call these inside the transaction (as we did pre 3.15)? It should be possible to run an async loop inside the single thread of the thread pool that only performes this coroutine (opening one addtional db connection).

Reply:
I tried calling them inside the transaction using async_to_sync() and it did not work as expected. RPM sync would always get an error in the _post_save() saying that the content was not yet in the database.

Reply:
Because it is a new connection and therefore not _in_ the transaction. XD
So this is in a really bad shape. As long as we still have the async thingies. But what more could we do...